rfd/0129-discovery-name-templating.md

```diff
@@ -0,0 +1,369 @@
+`^[a-z]([-a-z0-9]*[a-z0-9])?$`,
+and does not violate our kube cluster name validation regex `^[a-zA-Z0-9._-]+$`.
+
+#### GCP Discovery
+
+GCP discovery currently supports discovering only GKE kube clusters.
+
+GKE cluster names are unique within the same GCP project ID and location/zone.
+
+The discovery naming convention for GKE clusters shall be to append a suffix to
+the cluster name that includes:
+
+- Name of the Teleport GCP matcher type
+  - `gke`
+- GCP project ID
+  - These can be custom, but will only consist of characters, digits, hyphens.
+- GCP location
+
+```yaml
+    gcp:
+    - types: ["gke"]
+      locations: ["us-west1", "us-west2"]
+      tags:
+        "*": "*"
+      project_ids: ["my-project"]
+```
+
+If the discovery service is configured like the above, the discovery agent will
+discover GCP GKE kube clusters in "my-project" in the `us-west1` and `us-west2`
+locations.
+
+Now suppose GKE clusters named `foo` exist in each region.
+If the discovery service applies the new naming convention, the discovered
+resources should be named:
+
+- `foo-gke-us-west1-my-project`
+- `foo-gke-us-west2-my-project`
+
+This naming convention avoids name collisions between GKE clusters and does not
+collide with discovered AWS/Azure clusters.
+
+This naming convention does not violate our kube cluster name validation regex:
+`^[a-zA-Z0-9._-]+$`
+
+### `tsh` UX
+
+Users will be frustrated if they are forced to type out verbose resource names
+when using `tsh`.
+To avoid this poor UX, sub-commands should support prefix resource names or label
+matching to identify resources.
+
+The same UX should apply to all `tsh` sub-commands that take a resource name
+argument. These commands shall support
+`tsh <sub-command> [name | prefix] [key1=value1,key2=value2,...]` syntax:
+
+- `tsh db login`
+- `tsh db connect`
+- `tsh db env`
+- `tsh db config`
+- `tsh proxy db`
+- `tsh proxy kube`
+- `tsh kube login`
+
+#### `tsh` examples
+
+To illustrate the new UX for `tsh` sub-commands, here is an example using
+`tsh db connect` to select a database (the same applies for other commands):
+
+```sh
+$ tsh db ls
+Name   Description         Allowed Users       Labels                      Connect 
+------ ------------------- ------------------- --------------------------- ------- 
+foo-rds-us-west-1-0123456789012 RDS instance in ... [*] account-id=0123456789012,region=us-west-1,env=prod,...
+bar-rds-us-west-1-0123456789012 RDS instance in ... [*] account-id=0123456789012,region=us-west-1,env=dev,...
+bar-rds-us-west-2-0123456789012 RDS instance in ... [*] account-id=0123456789012,region=us-west-2,env=dev,...
+
+# connect by prefix name
+$ tsh db connect --db-user=alice --db-name-postgres foo
+#...connects to "foo-rds-us-west-1-0123456789012" by prefix...
+
+# ambiguous prefix name is an error
+$ tsh db connect --db-user=alice --db-name-postgres bar
+error: ambiguous database name could match multiple databases:
```

Comment:
let's make it:

Multiple databases match the name "bar". Please specify the full database name. You can find it by running `tsh db ls -v`

Reply:
initially, I thought copy/paste examples would be a good idea, but I'm not so sure anymore - the commands don't include the flags the user used previously. As a user I think i'm more likely to press the up arrow instead to modify my last command than copy/paste a new one. Also, printing example commands is bug prone, and it's frustrating to copy/paste a suggested command that doesn't work or doesn't do what I intended.

I prefer the error output from `tsh ssh` that just provides hints about how to resolve the ambiguity, with a generic example `Hint: try addressing the node by unique id (ex: tsh ssh user@node-id)`